src/main/groovy/com/morpheusdata/commvault/datasets/CommvaultStoragePoliciesDatasetProvider.groovy

```diff
@@ -0,0 +1,139 @@
+    public static final providerKey = 'commvaultStoragePolicies'
+    public static final providerDescription = 'A set of data from commvault storage policies'
+
+    CommvaultStoragePoliciesDatasetProvider(Plugin plugin, MorpheusContext morpheus) {
+        this.plugin = plugin
+        this.morpheusContext = morpheus
+    }
+
+    @Override
+    DatasetInfo getInfo() {
+        new DatasetInfo(
+                name: providerName,
+                namespace: providerNamespace,
+                key: providerKey,
+                description: providerDescription
+        )
+    }
+
+    @Override
+    Class<ReferenceData> getItemType() {
+        return ReferenceData.class
+    }
+
+    @Override
+    Observable list(DatasetQuery datasetQuery) {
+        log.debug("list: ${datasetQuery.parameters}")
+        def tmpAccount = datasetQuery.user.account
+        Cloud cloud
+        BackupProvider backupProvider
+        Long cloudId = datasetQuery.get("zoneId")?.toLong()
+        if(!cloud && cloudId) {
+            cloud = morpheusContext.services.cloud.get(cloudId)
+        }
+        if(cloud && cloud.backupProvider) {
+            backupProvider = morpheusContext.services.backupProvider.find(new DataQuery().withFilter("account", tmpAccount)
+                    .withFilter("id", cloud.backupProvider.id))
+        }
+        if(cloudId) {
+            return morpheusContext.services.referenceData.list(new DataQuery()
+                    .withFilter("category", "${backupProvider?.type?.code}.backup.storagePolicy.${backupProvider?.id}"))
+        }
+        return Observable.empty()
+    }
+
+    @Override
+    Observable<Map> listOptions(DatasetQuery datasetQuery) {
+        log.debug("listOptions: ${datasetQuery.parameters}")
+        def storagePolicies = []
+        def tmpAccount = datasetQuery.user.account
+        Cloud cloud
+        BackupProvider backupProvider
+        Long cloudId = datasetQuery.get("zoneId")?.toLong()
+        Long containerId = datasetQuery.get("containerId")?.toLong()
+        if(containerId && !cloudId) {
+            Workload workload = morpheusContext.services.workload.find(new DataQuery().withFilter("account", tmpAccount)
+                                                            .withFilter("containerId", containerId))
+            cloud = workload?.server?.cloud
+        }
+        if(!cloud && cloudId) {
+            cloud = morpheusContext.services.cloud.get(cloudId)
+        }
+        if(cloud && cloud.backupProvider) {
+            backupProvider = morpheusContext.services.backupProvider.find(new DataQuery().withFilter("account", cloud.account)
+                    .withFilter("id", cloud.backupProvider.id))
+        }
+
+        if(backupProvider) {
+            def results = morpheusContext.services.referenceData.list(new DataQuery()
+                    .withFilter("category", "${backupProvider?.type?.code}.backup.storagePolicy.${backupProvider?.id}"))
+            if (results.size() > 0) {
```

Comment:
Everything before this should be done in `list()` and then converted to a map in `listOptions()`:
```
list(datasetQuery).toList().blockingGet().collect {refData -> [name: refData.name, id: refData.id, value: refData.id] }
```

Reply:
@ddeyoung changes has been incorporated as suggested.